### PR TITLE
Implement observe mode

### PR DESCRIPTION
**Pull request by satyamsi**

### CODE REVIEW

File: enforcer/connection/connection.go [hunk 1]
```diff
@@ -142,75 +136,33 @@ func (c *TCPConnection) GetState() TCPFlowState {
...
-	}
 	// Logging information
-	if c.flowReported == 0 && len(c.logs) > 1 {
+	if c.flowReported == 0 {
 		zap.L().Error("Connection not reported",
-			zap.String("connection", c.String()),
-			zap.String("logs", logStr))
+			zap.String("connection", c.String()))
 	} else {
 		zap.L().Debug("Connection reported",
```
Comment (primalmotion) [inline enforcer/connection/connection.go]: do we really need the debug message here?

File: enforcer/connection/connection.go [hunk 2]
```diff
@@ -255,7 +206,7 @@ func (c *ProxyConnection) SetState(state ProxyConnState) {
 	c.state = state
 }
 
-//SetReported -- Set the flag to reported when the conn is reported
+// SetReported Sets the flag to reported when the conn is reported
```
Comment (primalmotion) [inline enforcer/connection/connection.go]: `SetReported sets the flag to reported when the conn is reported.`

File: enforcer/datapath/proxy/tcp/tcp.go [hunk 3]
```diff
@@ -460,51 +464,47 @@ L:
...
 					return errors.New("dropping because of reject rule on receiver")
 				}
 				conn.SetState(connection.ClientSendSignedPair)
 
 			case connection.ClientSendSignedPair:
-				token, err := p.tokenaccessor.CreateAckPacketToken(puContext.(*pucontext.PUContext), &conn.Auth)
+				token, err := p.tokenaccessor.CreateAckPacketToken(puContext, &conn.Auth)
 				if err != nil {
-					zap.L().Error("Failed to create ack token", zap.Error(err))
+					return fmt.Errorf("Failed to create ack token: %s", err)
```
Comment (primalmotion) [inline enforcer/datapath/proxy/tcp/tcp.go]: `unable to create ack token: %s`

File: enforcer/datapath/proxy/tcp/tcp.go [hunk 4]
```diff
@@ -460,51 +464,47 @@ L:
...
 				}
-				if serr := syscall.Sendto(downConn, token, 0, toAddr); serr != nil {
-					zap.L().Error("Sendto failed", zap.Error(serr))
-					return serr
+				if err := syscall.Sendto(downConn, token, 0, toAddr); err != nil {
+					return fmt.Errorf("Ack Send to failed: %s", err)
```
Comment (primalmotion) [inline enforcer/datapath/proxy/tcp/tcp.go]: `unable to send ack: %s`

File: enforcer/datapath/proxy/tcp/tcp.go [hunk 5]
```diff
@@ -559,18 +559,17 @@ E:
 					return err
 				}
 				claims.T.AppendKeyValue(enforcerconstants.PortNumberLabelString, strconv.Itoa(int(backendport)))
-				if index, plc := puContext.(*pucontext.PUContext).RejectRcvRules.Search(claims.T); index >= 0 {
-					zap.L().Error("Connection Dropped", zap.String("Policy ID", plc.(*policy.FlowPolicy).PolicyID))
-					p.reportRejectedFlow(flowProperties, conn, collector.DefaultEndPoint, puContext.(*pucontext.PUContext).ManagementID, puContext.(*pucontext.PUContext), collector.PolicyDrop, plc.(*policy.FlowPolicy))
+				index, action := puContext.(*pucontext.PUContext).SearchRejectRcvRules(claims.T)
+				if index >= 0 {
+					zap.L().Error("Connection Dropped", zap.String("Policy ID", action.PolicyID))
```
Comment (primalmotion) [inline enforcer/datapath/proxy/tcp/tcp.go]: hum, is that really an error?

File: enforcer/pucontext/pucontext.go [hunk 6]
```diff
@@ -13,25 +13,132 @@ import (
...
-	Extension         interface{}
-	IP                string
-	Mark              string
-	ProxyPort         string
-	Ports             []string
-	PUType            constants.PUType
-	SynToken          []byte
-	SynServiceContext []byte
-	SynExpiration     time.Time
+	ID                    string
```
Comment (primalmotion) [inline enforcer/pucontext/pucontext.go]: can you separate the public fields (first) and the private ones by a space?

File: enforcer/pucontext/pucontext.go [hunk 7]
```diff
@@ -13,25 +13,132 @@ import (
...
+				acceptRules.AddPolicy(rule)
+			}
+		} else if rule.Policy.Action.Rejected() {
+			if rule.Policy.Action.Observed() {
+				rejectObserveRules.AddPolicy(rule)
+			} else {
+				rejectRules.AddPolicy(rule)
+			}
+		} else {
+			continue
```
Comment (primalmotion) [inline enforcer/pucontext/pucontext.go]: if we reach here, should it be a panic? how can a policy be neither accepted nor rejected?

File: enforcer/connection/connection.go (hunk 1 above)
Comment (satyamsi) [reply enforcer/connection/connection.go]: yes. trying to debug duplicate reporting.

**Pull request merged**